File: 2017/day22.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day22.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile

grid = parsefile(file_name, [list, "\n"])

width = len(grid[0])
height = len(grid)

infected_points = set()
for y in range(height):
	for x in range(width):
		if grid[y][x] == "#":
			infected_points.add((x, y))


def part1():
	inf_points = infected_points.copy()
	pos = [width // 2, height // 2]
	inf_cnt = 0
	dir = 0  # up

	for i in range(10000):
		if (pos[0], pos[1]) in inf_points:  # turn right
			dir += 1
			if dir == 4:
				dir = 0

			# clean
			inf_points.remove((pos[0], pos[1]))
		else:  # turn left
			dir -= 1
			if dir == -1:
				dir = 3

			# infect
			inf_points.add((pos[0], pos[1]))
			inf_cnt += 1

		# move
		if dir == 0:  # up
			pos[1] -= 1
		elif dir == 2:  # down
			pos[1] += 1
		elif dir == 1:  # right
			pos[0] += 1
		elif dir == 3:  # left
			pos[0] -= 1

	return inf_cnt


def part2():
	inf_points = {p: 2 for p in infected_points}

	pos = [width // 2, height // 2]
	inf_cnt = 0
	dir = 0  # up

	for i in range(10000000):
		if (pos[0], pos[1]) not in inf_points:
			inf_points[(pos[0], pos[1])] = 0  # clean
		state = inf_points[(pos[0], pos[1])]
		if state == 0:  # clean, turn left
			dir -= 1
			if dir == -1:
				dir = 3

			# clean -> weakened
			inf_points[(pos[0], pos[1])] = 1
		elif state == 1:  # weakened, carry on
			# weakened -> infected
			inf_points[(pos[0], pos[1])] = 2

			inf_cnt += 1
		elif state == 2:  # infected, turn right
			dir += 1
			if dir == 4:
				dir = 0

			# infected -> flagged
			inf_points[(pos[0], pos[1])] = 3
		elif state == 3:  # flagged, reverse dir
			dir -= 2
			if dir < 0:
				dir += 4

			# flagged -> clean
			inf_points[(pos[0], pos[1])] = 0

		# move
		if dir == 0:  # up
			pos[1] -= 1
		elif dir == 2:  # down
			pos[1] += 1
		elif dir == 1:  # right
			pos[0] += 1
		elif dir == 3:  # left
			pos[0] -= 1

	# print(inf_points, pos)

	return inf_cnt


p1()
p2()

File: cpp/2017/day22.cpp
#include "../aocHelper.h"

class Day22 : public BaseDay
{
public:
	Day22() : BaseDay("22") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		unordered_set<uint32_t> points;

		int width;
		int height;

		// parse input

		int x = 0;
		int y = 0;
		while (*input != '\0')
		{
			if (*input == '\n')
			{
				width = x;
				y++;
				x = 0;
			}
			else
			{
				if (*input == '#')
				{
					points.insert((uint32_t) (x + 30'000) << 16 | (uint32_t) (y + 30'000));
				}
				x++;
			}
			input++;
		}

		height = y;

		// part 1
		unordered_set<uint32_t> p1_inf_points = points;
		pair<uint32_t, uint32_t> p1_pos = { (width / 2) + 30'000, (height / 2) + 30'000 };
		int p1_dir = 0; // up

		for (int i = 0; i < 10'000; i++)
		{
			uint32_t pos = p1_pos.first << 16 | p1_pos.second;
			if (p1_inf_points.contains(pos)) // turn right
			{
				p1_dir++;
				if (p1_dir == 4)
				{
					p1_dir = 0;
				}

				// clean
				p1_inf_points.erase(pos);
			}
			else // turn left
			{
				p1_dir--;
				if (p1_dir == -1)
				{
					p1_dir = 3;
				}

				// infect
				p1_inf_points.insert(pos);
				part1++;
			}

			// move
			switch (p1_dir)
			{
				case 0: // up
				{
					p1_pos.second--;
					break;
				}
				case 1: // right
				{
					p1_pos.first++;
					break;
				}
				case 2: // down
				{
					p1_pos.second++;
					break;
				}
				case 3: // left
				{
					p1_pos.first--;
					break;
				}
			}
		}

		// part 2
		unordered_map<uint32_t, int> p2_inf_points;
		for (auto& p : points)
		{
			p2_inf_points.insert({ p, 2 });
		}
		pair<uint32_t, uint32_t> p2_pos = { (width / 2) + 30'000, (height / 2) + 30'000 };
		int p2_dir = 0; // up

		for (int i = 0; i < 10'000'000; i++)
		{
			uint32_t pos = p2_pos.first << 16 | p2_pos.second;
			if (!p2_inf_points.contains(pos))
			{
				p2_inf_points.insert({ pos, 0 }); // clean
			}

			int state = p2_inf_points[pos];

			switch (state)
			{
				case 0: // clean, turn left
				{
					p2_dir--;
					if (p2_dir == -1)
					{
						p2_dir = 3;
					}

					// clean -> weakened
					p2_inf_points[pos] = 1;

					break;
				}
				case 1: // weakened, carry on
				{
					// weakened -> infected
					p2_inf_points[pos] = 2;
					part2++;

					break;
				}
				case 2: // infected, turn right
				{
					p2_dir++;
					if (p2_dir == 4)
					{
						p2_dir = 0;
					}

					// infected -> flagged
					p2_inf_points[pos] = 3;

					break;
				}
				case 3: // flagged, reverse dir
				{
					p2_dir -= 2;
					if (p2_dir < 0)
					{
						p2_dir += 4;
					}

					// flagged -> clean
					p2_inf_points[pos] = 0;

					break;
				}
			}

			// move
			switch (p2_dir)
			{
				case 0: // up
				{
					p2_pos.second--;
					break;
				}
				case 1: // right
				{
					p2_pos.first++;
					break;
				}
				case 2: // down
				{
					p2_pos.second++;
					break;
				}
				case 3: // left
				{
					p2_pos.first--;
					break;
				}
			}
		}

		return { part1, part2 };
	}
};
